File: Java/73.set-matrix-zeroes.java
public class Solution {
    public void setZeroes(int[][] matrix) {
        if(matrix == null || matrix.length == 0)    return;
        int m = matrix.length, n = matrix[0].length;
        boolean row = false, col = false;
        for(int i=0;i<m;i++)
            for(int j=0;j<n;j++){
                if(matrix[i][j] == 0){
                    if(i == 0)
                        row = true;
                    else if(j == 0)
                        col = true;
                    else{
                        matrix[i][0] = 0;
                        matrix[0][j] = 0;
                    }
                }
            }
        for(int i=m-1;i>=0;i--)
            for(int j=n-1;j>=0;j--)
                if(i == 0 && row == true || j == 0 && col == true || matrix[0][j] == 0 || matrix[i][0] == 0 )
                    matrix[i][j] = 0;
        return;
    }
}

File: 73.set-matrix-zeroes.js
/*
 * @lc app=leetcode id=73 lang=javascript
 *
 * [73] Set Matrix Zeroes
 */

// @lc code=start
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var setZeroes = function (matrix) {
  let R = matrix.length;
  let C = matrix[0].length;
  let rows = new Set();
  let cols = new Set();

  // Essentially, we mark the rows and columns that are to be made zero
  for (let i = 0; i < R; i++) {
    for (let j = 0; j < C; j++) {
      if (matrix[i][j] == 0) {
        rows.add(i);
        cols.add(j);
      }
    }
  }

  // Iterate over the array once again and using the rows and cols sets, update the elements.
  for (let i = 0; i < R; i++) {
    for (let j = 0; j < C; j++) {
      if (rows.has(i) || cols.has(j)) {
        matrix[i][j] = 0;
      }
    }
  }
};

// @lc code=end
